apps/prairielearn/src/lib/group-update.js

```diff
@@ -1,336 +1,207 @@
-const error = require('@prairielearn/error');
 const sqldb = require('@prairielearn/postgres');
 
+const { createServerJob } = require('./server-jobs');
+
 const sql = sqldb.loadSqlEquiv(__filename);
 
 module.exports = {
   /**
    * Update groups from a CSV file.
    *
-   * @param {number} assessment_id - The assessment to update.
+   * @param {string} assessment_id - The assessment to update.
    * @param {object} csvFile - An object with keys {originalname, size, buffer}.
-   * @param {number} user_id - The current user performing the update.
-   * @param {number} authn_user_id - The current authenticated user.
-   * @param {function} callback - A callback(err, job_sequence_id) function.
+   * @param {string} user_id - The current user performing the update.
+   * @param {string} authn_user_id - The current authenticated user.
    */
-  uploadInstanceGroups(assessment_id, csvFile, user_id, authn_user_id, callback) {
+  async uploadInstanceGroups(assessment_id, csvFile, user_id, authn_user_id) {
     if (csvFile == null) {
-      return callback(new Error('No CSV file uploaded'));
+      throw new Error('No CSV file uploaded');
     }
-    let params = { assessment_id };
-    sqldb.queryOneRow(sql.select_assessment_info, params, function (err, result) {
-      if (ERR(err, callback)) return;
-      const assessmentLabel = result.rows[0].assessment_label;
-      const course_instance_id = result.rows[0].course_instance_id;
-      const course_id = result.rows[0].course_id;
-      const options = {
-        course_id: course_id,
-        course_instance_id: course_instance_id,
-        assessment_id: assessment_id,
-        user_id: user_id,
-        authn_user_id: authn_user_id,
-        type: 'upload_groups',
-        description: `Upload group settings for ${assessmentLabel}`,
-      };
-      //create a job page to display progress and error message
-      serverJobs.createJobSequence(options, function (err, job_sequence_id) {
-        if (ERR(err, callback)) return;
-        callback(null, job_sequence_id);
-        const jobOptions = {
-          course_id: course_id,
-          course_instance_id: course_instance_id,
-          assessment_id: assessment_id,
-          user_id: user_id,
-          authn_user_id: authn_user_id,
-          type: 'upload_groups',
-          description: 'Upload group settings for ' + assessmentLabel,
-          job_sequence_id: job_sequence_id,
-          last_in_sequence: true,
-        };
-        serverJobs.createJob(jobOptions, function (err, job) {
-          if (err) {
-            logger.error('Error in createJob()', err);
-            serverJobs.failJobSequence(job_sequence_id);
-            return;
-          }
-          //create a lock to prevent two instructors from changing the same group assessement at the same time
-          const lockName = 'grouping assessment_id ' + assessment_id;
-          job.verbose(`Trying lock ${lockName}`);
-          namedLocks.tryLock(lockName, (err, lock) => {
```

Comment:
I _believe_ this doesn't actually need a lock, as there's just a single sproc call at the end of this. I'd appreciate a second opinion on that though.

Reply:
> I _believe_ this doesn't actually need a lock, as there's just a single sproc call at the end of this. I'd appreciate a second opinion on that though.

I don't think the issue is the lock with itself, but with the other process (`autoGroups`). In essence, it's trying to avoid having two group-creation processes (upload and automatic) running at the same time. The lock is the same in both cases. If that is indeed the case, it probably makes sense to have some way to ensure both lock names are created in a single place (e.g., a helper function) to avoid problems should one of the processes change the name by mistake.

Reply:
Ah fair - as in, we don't want the manual assignment to do its database write in the middle of the auto-assignment one? I can add back the lock in that case, as well as centralize the determination of the lock name.

FWIW, this locking won't cover any manual group creation/assignment occurring at the same time, as that doesn't run under a lock. Might be worth revisiting that so that `SELECT FOR NO KEY UPDATE` some common row, maybe the corresponding `group_configs` row?

Reply:
We definitely want to use a row lock here (presumably on the assessment instance?) rather than a named lock. But if we have a bulk-create method then we should be careful to handle each assessment in a separate query so we don't have multiple row locks associated with a single query (which can result in deadlocks if another query uses a different lock order).

Reply:
Specifically, the `assessment_groups_update` sproc is an example of where we should not acquire a lot of row locks, but instead we should loop in JS and use a separate query per assessment instance (locking it first).

Reply:
I believe we can lock on the `group_configs` table, which will have exactly one row for each assessment with group work enabled. Does that sound right? In that case, since we aren't locking on specific assessment instances, we won't have to be careful with lock orders or anything.

Reply:
Locking the group config row for the current assessment makes sense to me. It may be a good idea (possibly in a future PR) to do that for any group change, to avoid race conditions between this operation and other conflicting operations (e.g., if a user joins a group while this process is taking place, they may end up in two groups).

Reply:
If we lock on the `group_configs` row here then, as Jonatan says, we'll need to also lock on that same `group_configs` row when (for example) a student updates their group membership for an individual group instance. I'd prefer to lock on individual assessment instance rows. That will make the bulk operations a bit slower (and we would want to loop over assessment instances in JS rather than doing a bulk update in SQL) but it will avoid contention if we have 1000 students all trying to get into groups at the same time.

Reply:
Assessment instances don't exist yet when groups are created...

Reply:
That's a valid objection 😂

Reply:
Do you two want to see row locks added to this PR? Or are we willing to accept the (already existing) possibility of race conditions for a little longer until we can rework all the groups locking in an isolated and focused PR?

Reply:
I'm ok with pushing this forward.

Reply:
I added locking back to `uploadInstanceGroups` in 4eeeb97. We can revisit the overall locking strategy in a future PR.